### PR TITLE
add support for METADATA 2.1 in wheels

### PR DESCRIPTION
**Pull request by ali5h**
`wheel` has dropped support for `metadata.json`. This change adds the
functionality to parse METADATA 2.1 file and return the same
dictionary as deprecated `metadata.json`.

### CODE REVIEW

File: rules_python/whl.py [hunk 1]
```diff
@@ -107,9 +112,43 @@ def expand(self, directory):
 
   # _parse_metadata parses METADATA files according to https://www.python.org/dev/peps/pep-0314/
   def _parse_metadata(self, content):
-    # TODO: handle fields other than just name
-    name_pattern = re.compile('Name: (.*)')
-    return { 'name': name_pattern.search(content).group(1) }
+    metadata = {}
+    pkg_info = email.parser.Parser().parse(content)
```
Comment (clintharrison) [inline rules_python/whl.py]: This is not compatible with both Python 2 and 3. Running this with 2 will fail when there are unicode characters in the METADATA file (Grab python_dateutil-2.7.5-py2.py3-none-any.whl from PyPI for an example), as the email parser package uses cStringIO, which is only compatible with ASCII strings.

Comment (shariat) [reply rules_python/whl.py]: StringIO is compatible with unicode, but not mixed unicode and ASCII. I tested the change with the wheel you mentioned and it works

Comment (shariat) [reply rules_python/whl.py]: the output is 
```
package(default_visibility = ["//visibility:public"])

load("test", "requirement")

py_library(
    name = "pkg",
    srcs = glob(["**/*.py"]),
    data = glob(["**/*"], exclude=["**/*.py", "**/* *", "BUILD", "WORKSPACE"]),
    # This makes this directory a top-level in the python import
    # search path for anything that depends on this.
    imports = ["."],
    deps = [requirement("six")],
)
```

Comment (clintharrison) [reply rules_python/whl.py]: Err, you're right. Sorry, my bad. I got mixed up in my Python versions -- the issue is actually with Python 3, as the `email.feedparser.BufferedSubFile` used under the hood by `email.parser.Parser` uses `StringIO`. `zipfile.ZipFile.open()` always reads as bytes, which causes an error like the following:

```
  File "/Users/clint/code/rules_python/rules_python/whl.py", line 74, in metadata
    return self._parse_metadata(f)
  File "/Users/clint/code/rules_python/rules_python/whl.py", line 116, in _parse_metadata
    pkg_info = email.parser.Parser().parse(content)
  File "/usr/local/Cellar/python/3.7.1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/email/parser.py", line 57, in parse
    feedparser.feed(data)
  File "/usr/local/Cellar/python/3.7.1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/email/feedparser.py", line 175, in feed
    self._input.push(data)
  File "/usr/local/Cellar/python/3.7.1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/email/feedparser.py", line 103, in push
    self._partial.write(data)
TypeError: string argument expected, got 'bytes'
```

**Comment by ali5h** [comment]
@brandjon this is basically #90. I did it without knowing that #90 exists. Although, this one has much less code changes

**Comment by pcn** [comment]
`rfc822` doesn't exist in python3, I'm not sure the move to using it is going to work long-term.

**Comment by ali5h** [comment]
@pcn I have one with `email` package, will push the changes

**Comment by pcn** [comment]
That'd be cool, thanks.

**Comment by cristifalcas** [comment]
Hi. Can we have some updates on this one please?

**Comment by mdvorsky** [comment]
Here's an alternative implementation, with some tests as well: https://github.com/bazelbuild/rules_python/pull/198

**Comment by ali5h** [comment]
I wrote these fetch rules for pip using pip-compile https://github.com/ali5h/rules_pip_lock

**Pull request closed without merging**